flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java

```diff
@@ -115,7 +117,7 @@ private static class DppDimSideChecker {
         private final RelNode relNode;
         private boolean hasFilter;
         private boolean hasPartitionedScan;
-        private final List<ContextResolvedTable> tables = new ArrayList<>();
+        private final Set<ContextResolvedTable> tables = new HashSet<>();
```

Comment:
OOM happens because of 
```
for (ContextResolvedTable thisTable : new ArrayList<>(tables)) {
    if (!thisTable.getIdentifier().equals(catalogTable.getIdentifier())) {
        tables.add(catalogTable);
    }
}
```

in `setTables` method. That is, `tables.add` is used without checking if `tables` already contains the `catalogTable`

Reply:
Why do we write traversal comparisons like that？
```
                boolean hasAdded = false;
                for (ContextResolvedTable thisTable : new ArrayList<>(tables)) {
                    if (thisTable.getIdentifier().equals(catalogTable.getIdentifier())) {
                        hasAdded = true;
                        break;
                    }
                }
                if (!hasAdded) {
                    tables.add(catalogTable);
                }
```